storage/innobase/sync/srw_lock.cc

```diff
@@ -392,14 +391,22 @@ template void ssux_lock_impl<false>::wr_wait(uint32_t);
 template<bool spinloop>
 void ssux_lock_impl<spinloop>::rd_wait()
 {
-  for (;;)
+  if (spinloop)
   {
-    writer.wr_lock();
-    bool acquired= rd_lock_try();
-    writer.wr_unlock();
-    if (acquired)
-      break;
+    const unsigned delay= srw_pause_delay();
+
+    for (auto spin= srv_n_spin_wait_rounds; spin; spin--)
+    {
+      srw_pause(delay);
+      if (rd_lock_try())
+        return;
+    }
   }
+
+  writer.wr_lock();
+  ut_d(uint32_t lk=) readers.fetch_add(1, std::memory_order_relaxed);
+  writer.wr_unlock();
+  ut_ad(!(lk & WRITER));
```

Comment:
We could try adding a ‘pre-filtering’ step before these lines, to avoid unnecessarily locking `writer`:
```c++
#ifndef SUX_LOCK_GENERIC
  while (writer.lock.load(std::memory_order_acquire) & HOLDER)
  {
    writer.wait(lk);
    if (rd_lock_try())
      return;
  }
  if (rd_lock_try())
    return;
#endif
```

Reply:
It is not this simple. Before invoking `wait()` we would have to inform the holder that we are waiting, so that it would know to wake us once it is releasing the exclusive lock. 9430e397e0e83697d7b987ff8dfc8348b8dcf707 implements this, and it actually also works with `SUX_LOCK_GENERIC`.

Reply:
In 9430e397e0e83697d7b987ff8dfc8348b8dcf707 the `writer.wr_lock()` seems to be virtually unreachable during tests that I conducted. As far as I can tell, it is reachable with a tiny probability. We can remove this unnecessary contention point by implementing an infinite loop, even if the `HOLDER` flag on `writer.lock` is intermittently clear.

Reply:
0ba2fbe07a8e5d5fb59015cd137e27ce39533814 addresses this.